Optimize many? and one? using exists?-like query.

marshall-lee opened the pull request:
02da8aea made it possible to express `empty?` (and thus `any?`) in terms of `exists?`.
And now I see the potential of `exists?` algorithm to be extended to serve `many?` and `one?` methods too.

The idea is the following:

- `any?` runs `SELECT 1 ... LIMIT 1` and checks that result set is of size 1 (already implemented as `exists?`)
- `many?` runs `SELECT 1 ... LIMIT 2` and checks that result set is of size 2.
- `one?` runs `SELECT 1 ... LIMIT 2` and checks that result set is of size 1.

palkan commented on activerecord/lib/active_record/relation.rb:
```diff
@@ -292,13 +292,13 @@ def any?
     def one?
       return super if block_given?
-      limit_value ? records.one? : size == 1
+      limit_value ? records.one? : (loaded ? @records.length == 1 : exists_some?(max: 1))
```
Why not `@records.one?`? The same goes for `many?`

palkan replied:
And, I think, you can use `records` method instead of instance var here.

gmcgibbon reviewed (commented):
Still relevant in Rails 6 (11581afc01927565d57dd0e0156e983f92f8c2e0). The current implementation of `#one?` and `#many?` runs a `COUNT(*)` so I think this makes sense 👍

kaspth commented on activerecord/lib/active_record/relation.rb:
```diff
@@ -292,13 +292,13 @@ def any?
     def many?
       return super if block_given?
-      limit_value ? records.many? : size > 1
+      limit_value ? records.many? : (loaded ? @records.length > 1 : exists_some?(min: 2))
```
Agreed that the double ternary is tough to grasp, but don't find the flow-breaking double returns much better. `case` reads like the better deal:

```ruby
def many?
  case
  when block_given?
    super
  when limit_value || loaded?
    records.many?
  else
    exists_some?(min: 2)
  end
end
```

marshall-lee replied:
@kaspth I decided to use `if ... elsif ... else` chain.

marshall-lee committed:
Optimize many? and one? using exists?-like query.

02da8ae made it possible to express `empty?` (and thus `any?`) in terms of `exists?`.
And now I see the potential of `exists?` algorithm to be extended to serve `many?` and `one?` methods too.

The idea is the following:

- `any?` runs `SELECT 1 ... LIMIT 1` and checks that result set is of size 1 (already implemented as `exists?`).
- `many?` runs `SELECT 1 ... LIMIT 2` and checks that result set is of size 2.
- `one?` runs `SELECT 1 ... LIMIT 2` and checks that result set is of size 1.

marshall-lee commented on activerecord/lib/active_record/relation/finder_methods.rb:
```diff
@@ -316,12 +316,16 @@ def exists?(conditions = :none)
+      exists_some?(conditions: conditions)
+    end
+
+    def exists_some?(conditions: :none, min: 1, max: nil)
```
No, it's not possible. `min` and `max` are necessary further. I'd try to unfold the underyling logic:

1. `relations.exists?`: `min = 1`, `max = nil` (`nil` means *no bound*). So the condition would be something like:
```ruby
relation.limit(1).pluck(1).length == 1
```
2. `relations.many?`: `min = 2`, `max = nil`. So the condition would be something like:
```ruby
relation.limit(2).pluck(1).length == 2
```
3. `relations.one?`: `min = 1`, `max = 1`. So the condition would be something like:
```ruby
relation.limit(2).pluck(1).length == 1
```

So the above logic could be generalized by:

```ruby
size = relation.limit(max ? max + 1 : min).pluck(1).length
size >= min && (!max || size <= max)
```

marshall-lee replied:
So, it won't be possible to build `size >= min && size <= max` condition knowing only `limit`.

gmcgibbon replied:
It is possible, it just doesn't make sense given `#exists_some?`. If you renamed the method to `#exists_count` and moved the count comparisons to `#one?` and `#many?` respectively, it would cut down on some of the complexity. That was just an idea though, both approaches are valid.

marshall-lee commented:
@kaspth @gmcgibbon 

I rebased my branch, fixed conflicts and also did some changes:

- Made `exists_some?` protected
- Refactored a condition chain in `many?` and `one?` methods

marshall-lee commented:
Oh and finally tests are green

gmcgibbon commented on activerecord/test/cases/relations_test.rb:
```diff
@@ -1043,7 +1043,9 @@ def test_any
 
     assert_queries(3) do
-      assert posts.any? # Uses COUNT()
+      assert_sql(/SELECT\ +1/) do
```
It wasn't immediately clear what this SQL assertions is testing for. I would consider using something like `/SELECT  1 AS one/` instead.

marshall-lee replied:
Because for some reason
```sql
SELECT  1 AS one
```
is generated instead of
```sql
SELECT 1 AS one
```
I wasn't sure why is this extra space here. This is a quirk of `connection.select_values` so I thought it'd be better not to conform to it.

marshall-lee replied:
I simplified all the regexes with extra-space ones.

marshall-lee replied:
For some reason, Rails adds extra space after `SELECT` if there's `LIMIT`:
```ruby
[17] pry(main)> puts Post.all.select(1).to_sql
SELECT 1 FROM "posts"
=> nil
[18] pry(main)> puts Post.all.select(1).limit(1).to_sql
SELECT  1 FROM "posts" LIMIT 1
```

marshall-lee replied:
Let stick with 
```ruby
assert_sql(/SELECT  1 AS one/) do
...
end
```
until the Arel is fixed.

marshall-lee replied:
I tried to fix the double-space issue in #33977

gmcgibbon replied:
Nice work! Rebase and I think this can be merged.

marshall-lee replied:
#33977 was merged so I finally can rewrite these tests with single space :)

gmcgibbon commented on activerecord/test/cases/relations_test.rb:
```diff
@@ -1057,25 +1059,40 @@ def test_many
 
     assert_predicate posts, :many?
-    assert_not_predicate posts.limit(1), :many?
+    assert_sql(/SELECT\ +.posts.\.\*/) do
```
Same here, is there a reason why we can't do `/SELECT  "posts"\.\*/`?

marshall-lee replied:
@gmcgibbon PostgreSQL and MySQL have different quotation. In MySQL it'd be SELECT \`posts\`.*

matthewd commented:
Do we have some stats on real-world usage of these methods? I'm pretty dubious that the per-call gain wins out over the need for several calls when doing `if many?; elsif one?; else; end`.

And does this lose caching for multiple calls to (e.g.) `one?` on a single relation?

marshall-lee commented:
@matthewd 

> Do we have some stats on real-world usage of these methods?

Then I'd provide some execution plans from some of my real-world tables:

```
DATABASE=> explain analyze select 1 from posts where author_id = 43 limit 2;
 Limit  (cost=0.11..0.39 rows=2 width=4) (actual time=0.031..0.033 rows=2 loops=1)
   ->  Index Only Scan using index_posts_on_author_id on posts  (cost=0.11..169472.56 rows=1220058 width=4) (actual time=0.031..0.032 rows=2 loops=1)
         Index Cond: (author_id = 43)
         Heap Fetches: 2
 Planning time: 0.117 ms
 Execution time: 0.046 ms
(6 rows)

DATABASE=> explain analyze select 1 from posts where author_id = 43 limit 1;
 Limit  (cost=0.11..0.25 rows=1 width=4) (actual time=0.029..0.029 rows=1 loops=1)
   ->  Index Only Scan using index_posts_on_author_id on posts  (cost=0.11..169472.56 rows=1220058 width=4) (actual time=0.028..0.028 rows=1 loops=1)
         Index Cond: (author_id = 43)
         Heap Fetches: 1
 Planning time: 0.112 ms
 Execution time: 0.042 ms
(6 rows)

DATABASE=> explain analyze select count(*) from posts where author_id = 43;
 Finalize Aggregate  (cost=167878.34..167878.35 rows=1 width=8) (actual time=321.184..321.184 rows=1 loops=1)
   ->  Gather  (cost=167877.94..167878.34 rows=4 width=8) (actual time=321.150..321.180 rows=5 loops=1)
         Workers Planned: 4
         Workers Launched: 4
         ->  Partial Aggregate  (cost=166877.94..166877.94 rows=1 width=8) (actual time=315.441..315.442 rows=1 loops=5)
               ->  Parallel Index Only Scan using index_posts_on_author_id on posts  (cost=0.11..166725.43 rows=305014 width=0) (actual time=0.085..299.282 rows=212444 loops=5)
                     Index Cond: (author_id = 43)
                     Heap Fetches: 21160
 Planning time: 0.119 ms
 Execution time: 346.833 ms
(10 rows)
```

So even the subsequent calls of `many? ... one?` give a huge benefit over the one `count` call.

> And does this lose caching for multiple calls to (e.g.) one? on a single relation?

It does not lose cache. The actual query is being executed inside `skip_query_cache_if_necessary { ... }` block. In this sense it does not differ from previous `SELECT count(*)` solution — `count(:all)` is also being executed inside `skip_query_cache_if_necessary`: 

https://github.com/rails/rails/blob/068fe7dc9045856b822833db5cb7cb690e6000d7/activerecord/lib/active_record/relation/calculations.rb#L300

And current implementation of `exists?` is cacheable too:

https://github.com/rails/rails/blob/82d184839e573b2085d3ea8007f81ab31827edd0/activerecord/lib/active_record/relation/finder_methods.rb#L322

I just generalized the `exists?` approach.

matthewd commented:
Sorry, I said caching, but I meant memoization :confused:

`size` memoizes, making further calls on any of these methods ~free; the query cache avoids the actual network traffic, but it still involves comparatively more work.

simi commented:
:information_source: this is still valid pull request (`many?` is still ineffective). `COUNT(*)` is really expensive operation in DB (mainly in PostgreSQL by design). Any chance to revive this? Going thru the history I see just some minor code comments, but the overall acceptance seems positive.

Ping me if any help (rebase, code update, test, ...) is needed.

Diff for queries time from our production database (PostgreSQL) for `many?` call:

1. Rails 6 query

```sql
EXPLAIN ANALYZE SELECT count(*) FROM sales_requests;

COLD (first call):
Execution Time: 59427.049 ms
AFTER DISK CACHE kicked in:
Execution Time: 6740.494 ms
```

:information_source: Started multiple workers in PostgreSQL (using parallelization) to scan index in parallel.


2. this PR

```sql
EXPLAIN ANALYZE SELECT 1 FROM sales_requests LIMIT 2;
COLD (first call):
Execution Time: 0.056 ms
```

:information_source: used just quick sequence scan (quick for small amount of results).

marshall-lee commented:
@simi i think i should re-open this as a new PR with rebased changes. I sill believe that this change is what we all really need.

simi commented:
@marshall-lee you can also rebase and force push here to keep the history in PR. That would be the best. Ping me if you need any help.
